Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+
+```text
+exp -> exp '*' exp (1)
+exp -> exp '+' exp (2)
+exp -> integer     (3)
+```
+
+As well as the following states and actions:
+
+<table>
+  <tr>
+    <td>State</td>
+    <td>Position</td>
+    <td>Action</td>
+  </tr>
+  <tr>
+    <td>A</td>
+    <td>
+      <pre>exp -> exp . '*' exp (rule 1)
+exp -> exp '*' exp . (rule 1)
+exp -> exp . '+' exp (rule 2)</pre>
+    </td>
+    <td>
+      <pre>'+'     shift, and enter state D (reduce using rule 1)
+'*'     shift, and enter state C (reduce using rule 1)
+%eof    reduce using rule 1</pre>
+    </td>
+  </tr>
+  <tr>
+    <td>B</td>
+    <td>
+      <pre>exp -> exp . '*' exp (rule 1)
+exp -> exp . '+' exp (rule 2)
+exp -> exp '+' exp . (rule 2)</pre>
+    </td>
+    <td>
+      <pre>'+'     shift, and enter state D (reduce using rule 2)
+'*'     shift, and enter state C (reduce using rule 2)
+%eof    reduce using rule 2</pre>
+    </td>
+  </tr>
+  <tr>
+    <td>C</td>
+    <td>
+      <pre>exp -> exp '*' . exp (rule 1)</pre>
+    </td>
+    <td>
+      <pre>integer shift, and enter state E
+exp     goto state A</pre>
+    </td>
+  </tr>
+  <tr>
+    <td>D</td>
+    <td>
+      <pre>exp -> exp '+' . exp (rule 2)</pre>
+    </td>
+    <td>
+      <pre>integer shift, and enter state E
+exp     goto state B</pre>
+    </td>
+  </tr>
+  <tr>
+    <td>E</td>
+    <td>
+      <pre>exp -> integer .     (rule 3)</pre>
+    </td>
+    <td>
+      <pre>'+'     reduce using rule 3
+'*'     reduce using rule 3
+%eof    reduce using rule 3</pre>
+    </td>
+  </tr>
+</table>
+
+n.b.: I omitted some states for brevity.
+
+Remember that the dot (`.`) here indicates where the parser stopped while consuming input.
+For example, the first position in state A means that the parser has just finished consuming the first expression and it's about to consume a star.
+
+State A happens after we have consumed a `*`, while state B happens after consuming a `+`.
+
+Suppose we use `1 + 2 * 3` as the input, and suppose that the parser has arrived at `1 + 2 . * 3` in state B (coming from state D).
+It may either _shift_ to consume the `*` and go to state C or _reduce_ to accept that it has finished consuming the addition.
+
+In other words, if it shifts, it will parse `1 + (2 * 3)`, giving priority to `*`, and if it reduces, it will parse `(1 + 2) * 3`, giving priority to `+`.
+So if we want `*` to bind tighter than `+`, we want to reduce.
+
+Likewise, what if we have `1 + 2 + 3` as the input?
+Supposing that we stopped at `1 + 2 . + 3`, we enter state B again (coming from state D).
+It may either be parsed as `(1 + 2) + 3` or `1 + (2 + 3)`.
+If we shift, we make `+` right-associative, resulting in `1 + (2 + 3)`.
+If we reduce, we make `+` left-associative, resulting in `(1 + 2) + 3`.
+
+It means that an LR parser can use shift/reduce conflicts to either shift or reduce, allowing the programming to specify precedences and associativities for operators.
+
+##### Dangling else
```

Comment:
Yep, I left it there because Appel's book also describes these two as fam cases in conflicts, and I think it makes for a nice example as well. But I left a note saying that this is no different from the operator problem.